File: RMStimeseries.m
%% File import
% This will read in all files from the specified folder. NOTE: if an ".mat"
% files exists in the folder, it will be included in the exported
% dataframe!
clear all; close all
if strcmp(computer, 'PCWIN64')
    cd('C:\Users\Alex\Dropbox\Catholic U\VR_EXO\Neurospec development\neurospec_dev_data\mansucript data')
    files = dir(['*.mat']);
else
    cd('/Users/alexbrunfeldt/Dropbox/Catholic U/VR_EXO/Neurospec development/neurospec_dev_data/mansucript data')
    files = dir(['*.mat']);
end

%% Main Loop
for k = 1:length(files)
    filename = files(k).name;
    [rmswin, rmswinLoc] =  rmsWindow(filename, 7); % right now only chooses data for target location = 7
    if strcmp(computer, 'PCWIN64')
        cd('C:\Users\Alex\Dropbox\Catholic U\Manuscripts\VR_EXO_healthies\Review\MATLABoutput');
        %dlmwrite('rmswinDF', rmswin, '-append', 'delimiter', ',', 'precision', '%.12f'); % Write dataframe to folder as comma delim file NOTE: this appends data, so you must delete unwanted (old) files before running this
        dlmwrite('rmswinDFtargetLoc', rmswinLoc, '-append', 'delimiter', ',', 'precision', '%.12f');
        cd('C:\Users\Alex\Dropbox\Catholic U\VR_EXO\Neurospec development\neurospec_dev_data\mansucript data')
    else
        cd('/Users/alexbrunfeldt/Dropbox/Catholic U/Manuscripts/VR_EXO_healthies/Review/MATLABoutput')
        %dlmwrite('rmswinDF', rmswin, '-append', 'delimiter', ',', 'precision', '%.12f'); % Write dataframe to folder as comma delim file NOTE: this appends data, so you must delete unwanted (old) files before running this
        dlmwrite('rmswinDFtargetLoc', rmswinLoc, '-append', 'delimiter', ',', 'precision', '%.12f');
        cd('/Users/alexbrunfeldt/Dropbox/Catholic U/VR_EXO/Neurospec development/neurospec_dev_data/mansucript data');
    end
end
%

%% RMS windowing function
function [rmswin, rmswinLoc] =  rmsWindow(filename, targetLoc)
%% Signal preparation
load(filename);
% Determine resting bias
ldelt_bias = mean(ldelt(1:100));
rdelt_bias = mean(rdelt(1:100));
lbicep_bias = mean(lbicep(1:100));
rbicep_bias = mean(rbicep(1:100));

% Remove resting bias (offset correct)
ldeltOC = ldelt - ldelt_bias;
rdeltOC = rdelt - rdelt_bias;
lbicepOC = lbicep - lbicep_bias;
rbicepOC = rbicep - rbicep_bias;

% Fix rare case where the onset for trial one is negative (ie: onsetoffsetLH/RH(1,1) = neg value)
if onsetoffsetLH(1,1) < 0
    onsetoffsetLH(1,1) = 201; % Changed to 201 here to compensate for epoch shift. See Lab book pg 114.
end
if onsetoffsetLH(1,2) < 0
    onsetoffsetLH(1,2) = onsetoffsetLH(1,1) + 100; % Even more rare occurance when offset for trial 1 is negative. In this case, make offset 100 sample after onset
end
if onsetoffsetRH(1,1) < 0
    onsetoffsetRH(1,1) = 201;
end
if onsetoffsetRH(1,2) < 0
    onsetoffsetRH(1,2) = onsetoffsetRH(1,1) + 100; % Even more rare occurance when offset for trial 1 is negative. In this case, make offset 100 sample after onset
end

% Fix ultra rare case where final offset value is outside bounds of data array
if onsetoffsetLH(end,2) > length(ldelt)
    onsetoffsetLH(end,2) = length(ldelt); % set final offset to end of data
end
if onsetoffsetRH(end,2) > length(rdelt)
    onsetoffsetRH(end,2) = length(rdelt); % set final offset to end of data
end
% Shift onset/offset markers forward by 200ms. This is because muscle
% activity begins ~200ms before movement onset, and we have a long tail of
% null activity after movement offset. See Lab book pg 114.
onsetoffsetLH(:,:) = onsetoffsetLH(:,:) - 200;
onsetoffsetRH(:,:) = onsetoffsetRH(:,:) - 200;
% Resample loop
nsamp = 5000;
for i = 1:length(targets)
    % epoch into individual trial timeseries
    tempLdelt = ldeltOC(onsetoffsetLH(i,1):onsetoffsetLH(i,2));
    tempRdelt = rdeltOC(onsetoffsetRH(i,1):onsetoffsetRH(i,2));
    tempLbicep = lbicepOC(onsetoffsetLH(i,1):onsetoffsetLH(i,2));
    tempRbicep = rbicepOC(onsetoffsetRH(i,1):onsetoffsetRH(i,2));
    
    % Resample to same length data,
    %then concat into single array (per muscle)
    xqLdelt = linspace(1, length(tempLdelt), nsamp); % Resample to this length array
    xqRdelt = linspace(1, length(tempRdelt), nsamp); % Resample to this length array
    ldeltrs(:,i) = interp1(tempLdelt, xqLdelt, 'linear', 'extrap')';
    rdeltrs(:,i) = interp1(tempRdelt, xqRdelt, 'linear', 'extrap')';
    
    xqLbicep = linspace(1, length(tempLbicep), nsamp); % Resample to this length array
    xqRbicep = linspace(1, length(tempRbicep), nsamp); % Resample to this length array
    lbiceprs(:,i) = interp1(tempLbicep, xqLbicep, 'linear', 'extrap')';
    rbiceprs(:,i) = interp1(tempRbicep, xqRbicep, 'linear', 'extrap')';
end
ldeltmean = mean(ldeltrs,2);
rdeltmean = mean(rdeltrs,2);
lbicepmean = mean(lbiceprs,2);
rbicepmean = mean(rbiceprs,2);

% figure;tiledlayout(4,1)
% nexttile; plot(ldeltmean,'b'); title('Left Deltoid')
% nexttile; plot(rdeltmean,'r'); title('Right Deltoid')
% nexttile; plot(lbicepmean,'b'); title('Left Bicep')
% nexttile; plot(rbicepmean,'r'); title('Right Bicep')
%% RMS windowing (see lab book 113)
winsize = 100; % window size; must be factor of 'nsamp'
lengthArray = nsamp/winsize;
% Detrend data
ldeltDT = detrend(ldeltrs);
rdeltDT = detrend(rdeltrs);
lbicepDT = detrend(lbiceprs);
rbicepDT = detrend(rbiceprs);

% %Filter data
% bpfreq = [5 250];
% filtLdelt = bandpass(ldeltDT,bpfreq,freq);
% filtRdelt = bandpass(rdeltDT,bpfreq,freq);
% filtLbicep = bandpass(lbicepDT,bpfreq,freq);
% filtRbicep = bandpass(rbicepDT,bpfreq,freq);

% RMSE calculation
% NOTE: for some reason, I'm getting LARGE edge effects on the filtering,
% so I am removing for now. This is OK because the window RMS calc, plus
% averaging across targets, acts as a large low-pass filter. Might still
% get some low frequency articfacts (e.g. movement artifacts)
for i = 1:lengthArray
    ldeltrms(i,:) = mean(ldeltDT((i-1)*winsize+1:i*winsize,:).^2,1);
    rdeltrms(i,:) = mean(rdeltDT((i-1)*winsize+1:i*winsize,:).^2,1);
    lbiceprms(i,:) = mean(lbicepDT((i-1)*winsize+1:i*winsize,:).^2,1);
    rbiceprms(i,:) = mean(rbicepDT((i-1)*winsize+1:i*winsize,:).^2,1);
end

%% Trial timeseries for target = 'targetLoc'
% creates an output of all 16 timeseries for the 16 reaches to the target
% specified in 'targetLoc'. This can be used in R for the baseline
% correction followed by SEM calculation (in R). See lab book pg. 114.
ldeltRMStargetLoc = ldeltrms(:,find(targets == targetLoc));
rdeltRMStargetLoc = rdeltrms(:,find(targets == targetLoc));
lbicepRMStargetLoc = lbiceprms(:,find(targets == targetLoc));
rbicepRMStargetLoc = rbiceprms(:,find(targets == targetLoc));
%These are used in next section to create the export DF
%
%% Take mean by target location
ldeltrmsmean = zeros(lengthArray,6);
rdeltrmsmean = zeros(lengthArray,6);
ldeltrmsstd = zeros(lengthArray,6);
rdeltrmsstd = zeros(lengthArray,6);
lbiceprmsmean = zeros(lengthArray,6);
rbiceprmsmean = zeros(lengthArray,6);
lbiceprmsstd = zeros(lengthArray,6);
rbiceprmsstd = zeros(lengthArray,6);
for i = 5:10 % targets number from 5 to 10
    % Means
    ldeltrmsmean(:,i-4) = mean(ldeltrms(:, find(targets == i)), 2);
    rdeltrmsmean(:,i-4) = mean(rdeltrms(:, find(targets == i)), 2);
    lbiceprmsmean(:,i-4) = mean(ldeltrms(:, find(targets == i)), 2);
    rbiceprmsmean(:,i-4) = mean(rdeltrms(:, find(targets == i)), 2);
    % Standard deviations
    ldeltrmsstd(:,i-4) = std(ldeltrms(:, find(targets == i)), 0, 2);
    rdeltrmsstd(:,i-4) = std(rdeltrms(:, find(targets == i)), 0, 2);
    lbiceprmsstd(:,i-4) = std(ldeltrms(:, find(targets == i)), 0, 2);
    rbiceprmsstd(:,i-4) = std(rdeltrms(:, find(targets == i)), 0, 2);
end

% concatinate the mean/std arrays
% first create the hand and muscle factors
handL = repelem(1,lengthArray)'; % left hand = 1
handR = repelem(2,lengthArray)'; % right hand = 2
muscleD = repelem(1,lengthArray)'; % Deltoid = 1
muscleB = repelem(2,lengthArray)'; % Bicep = 2
sample = linspace(1,lengthArray,lengthArray)';
% then concat into full data arrays
% NOTE: Here is where we choose a specific target - this was done to limit
% the need for additional factoring/processing either here or in R. This is
% in response to Reviewer #2 in the healthy control study, so its highly specific for this manuscript
tempDF = [handL muscleD sample ldeltrmsmean(:,targetLoc-4) ldeltrmsstd(:,targetLoc-4);...
          handR muscleD sample rdeltrmsmean(:,targetLoc-4) rdeltrmsstd(:,targetLoc-4); ...
          handL muscleB sample lbiceprmsmean(:,targetLoc-4) lbiceprmsstd(:,targetLoc-4); ...
          handR muscleB sample rbiceprmsmean(:,targetLoc-4) rbiceprmsstd(:,targetLoc-4)];

tempDFtargetLoc = [handL muscleD sample ldeltRMStargetLoc;...
          handR muscleD sample rdeltRMStargetLoc; ...
          handL muscleB sample lbicepRMStargetLoc; ...
          handR muscleB sample rbicepRMStargetLoc];


%% Add other factors (subID, cond, block, etc.)
formatIn = 'mm_dd_yy'; % Note, this is only unique when only one participant per day is run...
subID = str2num([num2str(datenum(filename(1:8),formatIn)) filename(10:11)]); % Convert mm_dd_yy of "file" into a datenum unique integer, then add participant order (if more than one person tested per day)
cond = filename(13:14);
switch cond
    case{'bl'}
        condition = 1;
    case{'rh'}
        condition = 2;
    case{'lh'}
        condition = 3;
    case{'bh'}
        condition = 4;
end

subid = ones(length(tempDF),1)*subID; % subid vector
cond = ones(length(tempDF),1)*condition; % condition vector as numeric
block = ones(length(tempDF),1)*str2num(filename(16:17)); % block (1-7 for controls, 1-3 for stroke)

%% Full dataframe for export!
rmswin = [subid cond block tempDF];
rmswinLoc = [subid cond block tempDFtargetLoc];
end

File: RMStimeseries.R
library(tidyverse)
library(readxl)
library(cowplot)
# Load in data---------------------------
# PC computer
setwd('C:\\Users\\Alex\\Dropbox\\Catholic U\\Manuscripts\\VR_EXO_healthies\\Review\\MATLABoutput') # PC
p.table = read_xlsx('C:\\Users\\Alex\\Dropbox\\Catholic U\\VR_EXO\\VR_participant_list.xlsx')
# 
# # Mac computer
# setwd('//Users//alexbrunfeldt//Dropbox//Catholic U//Manuscripts//VR_EXO_healthies//Review//MATLABoutput') # PC
# p.table = read_xlsx('//Users//alexbrunfeldt//Dropbox//Catholic U//VR_EXO//VR_participant_list.xlsx')

df.rms = read.delim('rmswinDF',header = FALSE, sep = ",", na.strings = 'NaN') # load in data
#
# Set column names; recode factors---------------
# Make sure the colnames match these (copied straight from Matlab)
# [subid cond block handL sample lhDmean(:,target-4) lhDstd(:,target-4)];
colnames(df.rms) = c('Subject', 'Condition', 'Block', 'Hand', 'Muscle', 'Sample',
                      'rms.mean', 'rms.std') # rename variables

# Recode Condition to match intervention
df.rms = df.rms %>% mutate(Condition = recode(Condition,
                                                '1' = "Baseline",
                                                '2' = "RH weighted",
                                                '3' = "LH weighted",
                                                '4' = "Combo"),
                           Hand = recode(Hand,
                                           '1' = "Left",
                                           '2' = "Right"),
                           Muscle = recode(Muscle,
                                           '1' = "Deltoid",
                                           '2' = "Bicep"))
#

# Set Load Factor -----------------------
# Identify in the 'Load' factor in the VR_participant_list excel file
subs.day1.30 = subset(p.table, grepl("3", Intervention))$subID01 # Find subjects who did 30% on day 1
subs.day1.60 = subset(p.table, grepl("6", Intervention))$subID01 # Find subjects who did 60% on day 1
subs.day2.30 = subset(p.table, !subID01 %in% subs.day1.30)$subID02 # Find subjects who did 30% on day 2
subs.day2.60 = subset(p.table, !subID01 %in% subs.day1.60)$subID02 # Find subjects who did 60% on day 2

# Add in the 'Load' and 'Day' factors
df.rms = df.rms %>% mutate(Load = case_when(Subject %in% subs.day1.30 ~ "30%",
                                              Subject %in% subs.day1.60 ~ "60%",
                                              Subject %in% subs.day2.30 ~ "30%",
                                              Subject %in% subs.day2.60 ~ "60%")) %>%
  mutate(Day = case_when(Subject %in% c(subs.day1.30, subs.day1.60) ~ "Day 1",
                         Subject %in% c(subs.day2.30, subs.day2.60) ~ "Day 2")) %>%
  relocate(Load, .after = Hand) %>% relocate(Day, .after = Load)

# Replace day 2 Subject identiers with the corresponding day 1 subject identifiers. NOTE: this is coded in the p.table variable from the Excel file!!!
for (i in 1:nrow(p.table)){
  if (!is.na(p.table$subID01[i]) & !is.na(p.table$subID02[i])) # Only recode non-"NA" values
  {
    df.rms[df.rms$Subject == p.table$subID02[i],]$Subject = p.table$subID01[i]
  }
}

# Remove outlier participants (ie: left handers, no-shows, etc...)
outliers = c(73823902)
df.rms = subset(df.rms, !Subject %in% outliers)
# Set Factors
factors = c('Subject', 'Condition', 'Hand', 'Muscle', 'Load', 'Day')
df.rms[,factors] = lapply(df.rms[,factors], factor)
rm(subs.day1.30, subs.day1.60, subs.day2.30, subs.day2.60, factors, i)

#

# Baseline correct --------------------------------------------------------
# Since raw EMG data are not internally normalized (different sensor configs, etc.), nor do I have MVC data; I need to baseline correct
# I'll do this by taking each of the 100 trials per muscle, hand, block as a % change from the mean at baseline.

# Baseline correct (% Change from Block 1)
df.rms = df.rms %>% group_by(Subject, Muscle, Hand, Day) %>% mutate(rms.mean.bc = 100*(rms.mean - mean(rms.mean[Block == 1]))/mean(rms.mean[Block == 1]),
                                                                    rms.std.bc = 100*(rms.std - mean(rms.std[Block == 1]))/mean(rms.std[Block == 1]))


# Plotting----------------------
ggplot(data = subset(df.rms, Condition == "RH weighted" & Load == "60%" & !Block %in% c(3,5) & Muscle == "Deltoid"), aes(x = Sample, y = rms.mean, color = Hand))+
  geom_point()+
  facet_wrap(~Subject)

ggplot(data = subset(df.rms, Load == "60%" & !Block %in% c(3,5)), aes(x = Sample, y = rms.mean, color = Hand))+
  stat_summary(fun.data = 'mean_se')+
  facet_grid(rows = vars(Condition))

# Plot sub 73824601, load 60; use errorbar = SEM
ggplot(data = subset(df.rms, Subject == "73824601" & Load == "60%" & !Block %in% c(1,3,5) & Muscle == "Deltoid"), aes(x = Sample, y = rms.mean, color = Hand))+
  geom_line(size = 1)+
  geom_ribbon(aes(ymin = rms.mean - rms.std/sqrt(16), ymax = rms.mean + rms.std/sqrt(16), fill = Hand), alpha = 0.3)+
  theme_cowplot()+
  panel_border()+
  facet_grid(rows = vars(Condition))+
  theme(text = element_text(size = 26),
        strip.text.x = element_text(size = 24),
        axis.text = element_text(size = 14),
        strip.background = element_blank())
